File: backend/src/main/java/com/tkpm/sms/entity/Student.java
package com.tkpm.sms.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.tkpm.sms.enums.Gender;
import jakarta.persistence.*;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.time.LocalDate;

@Entity
@Table(name = "students")
@Builder
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Student {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    String id;

    @Column(name = "student_Id", unique = true)
    String studentId;

    @Pattern(regexp = "^[\\p{L}\\s]*$", message = "Name must contain only letters and spaces")
    @NotNull
    String name;

    LocalDate dob;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    Gender gender;

    Integer course;

    @Email
    @NotNull
    @Column(unique = true)
    String email;

    @NotNull
    @Pattern(regexp = "^\\+\\d{1,3}\\d{6,12}$", message = "Phone number must be in international format (e.g., +84123456789)")
    @Column(unique = true)
    String phone;

    // many-to-one relationship with status
    @ManyToOne
    @JoinColumn(name = "status_id")
    Status status;

    // many-to-one relationship with program
    @ManyToOne
    @JoinColumn(name = "program_id")
    Program program;

    // many-to-one relationship with faculty
    @ManyToOne
    @JoinColumn(name = "faculty_id")
    Faculty faculty;

    @JsonBackReference
    @OneToOne(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "permanent_address_id")
    Address permanentAddress;

    @JsonBackReference
    @OneToOne(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "temporary_address_id")
    Address temporaryAddress;

    @JsonBackReference
    @OneToOne(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "mailing_address_id")
    Address mailingAddress;

    @OneToOne(cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonBackReference
    @JoinColumn(name = "identity_id")
    Identity identity;
}

File: frontend/src/types/student.ts
import { Address } from '@/types/address';
import IdentityDocument from './identityDocument';

// Using string types instead of enums since values come from the API
export type Gender = string;
export type Faculty = string;
export type Status = string;

export default interface Student {
  id: string;
  studentId: string;
  name: string;
  dob: Date;
  gender: Gender;
  faculty: Faculty;
  course: number;
  program: string;
  email: string;
  permanentAddress: Address;
  temporaryAddress: Address;
  mailingAddress: Address;
  phone: Phone;
  status: Status;
  identity: IdentityDocument;
}

export interface CreateStudentDTO {
  studentId: string;
  name: string;
  dob: Date;
  gender: Gender;
  faculty: Faculty;
  course: number;
  program: string;
  email: string;
  permanentAddress: Address;
  temporaryAddress: Address;
  mailingAddress: Address;
  phone: Phone;
  status?: Status;
  identity: IdentityDocument;
}

export interface UpdateStudentDTO {
  name?: string;
  dob?: Date;
  gender?: Gender;
  faculty?: Faculty;
  course?: number;
  program?: string;
  email?: string;
  permanentAddress: Address;
  temporaryAddress: Address;
  mailingAddress: Address;
  phone?: Phone;
  status?: Status;
  identity: IdentityDocument;
}

export interface Phone {
  phoneNumber: string;
  countryCode: string;
}

export const mapToStudent = (data: any): Student => {
  const res = {
    id: data.id,
    studentId: data.studentId,
    name: data.name,
    dob: data.dob ? new Date(data.dob) : new Date(),
    gender: data.gender,
    faculty: data.faculty,
    course: data.course,
    program: data.program,
    email: data.email,
    permanentAddress: data.permanentAddress,
    temporaryAddress: data.temporaryAddress,
    mailingAddress: data.mailingAddress,
    phone: data.phone,
    status: data.status,
    identity: data.identity,
  };

  return res;
};
